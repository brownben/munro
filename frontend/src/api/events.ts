import { getData, postData, putData, deleteData } from './requests'

export const getEvent = (id: string): Promise<LeagueEvent | null> =>
  getData<LeagueEvent>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Fetching Event Details',
  })

export const getEvents = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events',
    customErrorMessage: 'Problem Fetching Events',
  })

export const getLeagueEvents = (
  league: string,
  uploadKey?: boolean
): Promise<LeagueEvent[] | null> => {
  const apiLocation = uploadKey
    ? `/api/leagues/${league}/events/uploadKey`
    : `/api/leagues/${league}/events`

  return getData<LeagueEvent[]>({
    apiLocation,
    customErrorMessage: 'Problem Fetching Events',
  })
}

export const getLatestResults = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events/latest-results',
    customErrorMessage: 'Problem Fetching Events',
  })

export const createEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/events`,
    data,
    customErrorMessage: 'Problem Creating Event',
    customSuccessMessage: `Event \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/events/${data.id}`,
    data,
    customErrorMessage: 'Problem Updating Event',
    customSuccessMessage: `Event \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteEvent = (
  id: string,
  name: string
): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Deleting Event',
    customSuccessMessage: `Event \`${name}\` Deleted`,
  })
