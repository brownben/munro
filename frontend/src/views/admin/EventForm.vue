<template>
  <Layout
    :title="title"
    :not-found="$route.path.includes('/edit') && !loading && !event.id"
  >
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      :title="`Munro - ${title}`"
      description
    />

    <form class="col-span-2" @submit.prevent="submit">
      <TextInput v-model.trim="event.name" label="Name:" />
      <TextInput
        v-model.trim="event.date"
        label="Date: (YYYY-MM-DD)"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.organiser"
        label="Club/ Organiser:"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.website"
        label="Website: (URL)"
        type="url"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.results"
        label="Results: (URL)"
        type="url"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.winsplits"
        label="Winsplits: (URL)"
        type="url"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.routegadget"
        label="Routegadget: (URL)"
        type="url"
        class="mt-4"
      />
      <DropdownInput
        v-model="event.league"
        :list="leagues.map((league) => league.name)"
        label="League:"
        class="mt-4"
      />
      <TextInput
        v-model.trim="event.moreInformation"
        label="More Information:"
        class="mt-4"
      />
      <CheckboxInput
        v-model="event.userSubmittedResults"
        label="Allow Users to Submit Results"
        class="mt-6 text-left"
      />
      <button v-if="$route.path.includes('/edit')" class="mt-8 button-lg">
        Update Event
      </button>
      <button v-else class="mt-8 button-lg">Create Event</button>
    </form>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, watch, computed } from 'vue'
import { useRoute, useRouter } from 'vue-router'
import { useStore } from 'vuex'

import Layout from '../../components/Layout.vue'
import DropdownInput from '../../components/inputs/DropdownInput.vue'
import TextInput from '../../components/inputs/TextInput.vue'
import CheckboxInput from '../../components/inputs/CheckboxInput.vue'

import { toSingleString } from '../../scripts/typeHelpers'

import { getLeagues } from '../../api/leagues'
import {
  getEvent,
  createEvent as apiCreateEvent,
  updateEvent as apiUpdateEvent,
} from '../../api/events'

const store = useStore()
const router = useRouter()
const route = useRoute()

const loading = ref(true)
const leagues = ref<League[]>([])
const event = ref<LeagueEvent>({
  id: '',
  name: '',
  date: '',
  resultUploaded: false,
  organiser: '',
  moreInformation: '',
  website: '',
  results: '',
  winsplits: '',
  routegadget: '',
  userSubmittedResults: false,
  league: '',
})

const refreshDetails = async () => {
  const routeParamsId = toSingleString(route.params.id)

  getLeagues().then((data) => {
    leagues.value = data ?? []
  })

  if (routeParamsId) {
    loading.value = true
    await getEvent(routeParamsId).then((data) => {
      if (data) event.value = data
    })
    loading.value = false
  }
}

const validateForm = () => {
  if (event.value.name === '' || event.value.league === '') {
    store.dispatch(
      'createMessage',
      'Please Ensure Name and League Fields are not Blank'
    )
    return false
  } else if (
    event.value.name.includes('/') ||
    event.value.name.includes('\\')
  ) {
    store.dispatch(
      'createMessage',
      "Please Ensure Name doesn't Include any Slashes"
    )
    return false
  } else return true
}
const createEvent = () => {
  if (validateForm())
    return apiCreateEvent(event.value)
      .then(() => router.push(`/leagues/${event.value.league}`))
      .catch(() => false)
}
const updateEvent = () => {
  if (validateForm())
    return apiUpdateEvent(event.value)
      .then(() => router.push(`/leagues/${event.value.league}`))
      .catch(() => false)
}
const submit = () =>
  route.path.includes('/edit') ? updateEvent() : createEvent()

const title = computed(() =>
  route.path.includes('/edit') ? 'Edit Event' : 'Create Event'
)

watch(route, refreshDetails, { immediate: true })
</script>
