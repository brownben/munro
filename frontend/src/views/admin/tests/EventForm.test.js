/*
  Event Form View Unit Tests
*/

import axios from 'axios'
import { mount } from '@vue/test-utils'
import EventForm from '@/views/admin/EventForm.vue'
import { sampleSingleEvent } from '@/tests/test data/events'

jest.mock('axios')

beforeEach(() => {
  jest.resetAllMocks()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '', params: { league: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Creation Form - Renders Correctly', () => {
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '', params: { league: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeTruthy()
  expect(wrapper.element).toMatchSnapshot()
})

test('Update Form - Renders Correctly', () => {
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeFalsy()
  expect(wrapper.element).toMatchSnapshot()
})

test('Submit Method', () => {
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const getCreateSpy = jest.spyOn(wrapper.vm, 'createEvent')
  const getUpdateSpy = jest.spyOn(wrapper.vm, 'updateEvent')

  wrapper.setData({ create: true })
  wrapper.vm.submit()
  expect(getCreateSpy).toHaveBeenCalledTimes(1)
  expect(getUpdateSpy).toHaveBeenCalledTimes(0)

  jest.resetAllMocks()

  wrapper.setData({ create: false })
  wrapper.vm.submit()
  expect(getUpdateSpy).toHaveBeenCalledTimes(1)
  expect(getCreateSpy).toHaveBeenCalledTimes(0)
})

test('Get Leagues - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Leagues - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching List of Leagues'
  )
})

test('Get Event Details - Makes Request with Correct ID', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: 'eventID' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getEventDetails()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/events/eventID')
})

test('Get Event Details - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: {
        path: '/events/1/edit',
        params: { league: '', id: 'SprintelopeBushEstate2018-05-09' },
      },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: sampleSingleEvent })
  await wrapper.vm.getEventDetails()
  expect(wrapper.findAll('input').at(0).element.value).toBe(
    sampleSingleEvent.name
  )
  expect(wrapper.findAll('input').at(1).element.value).toBe(
    sampleSingleEvent.date
  )
  expect(wrapper.findAll('input').at(2).element.value).toBe(
    sampleSingleEvent.organiser
  )
  expect(wrapper.findAll('input').at(6).element.value).toBe(
    sampleSingleEvent.routegadget
  )
})

test('Get Event Details - No Event Found', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.notFound).toBeFalsy()
  axios.get.mockResolvedValue({ data: null })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.notFound).toBeTruthy()
})

test('Get Event Details - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getEventDetails()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Getting Event Details'
  )
})

test('Check Validation Works', () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })

  wrapper.setData({ name: '', league: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )

  wrapper.setData({ name: 'Value', league: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(2)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )

  wrapper.setData({ name: '', league: 'Value' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(3)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )

  wrapper.setData({ name: '/', league: 'A' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(4)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    "Please Ensure Name doesn't Include any Slashes"
  )

  wrapper.setData({ name: '\\', league: 'A' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(5)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    "Please Ensure Name doesn't Include any Slashes"
  )

  wrapper.setData({ name: 'Hello/ bye', league: 'A' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(6)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    "Please Ensure Name doesn't Include any Slashes"
  )

  wrapper.setData({ name: 'A Value', league: 'Another Value' })
  expect(wrapper.vm.validateForm()).toBeTruthy()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(6)
})

test('Return to League Home Page', () => {
  const mockAddMessageFunction = jest.fn()
  const mockRouterPushFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $router: { push: mockRouterPushFunction },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ league: 'Test League' })
  wrapper.vm.returnToLeaguePage({
    data: { message: 'This is the Message Returned' },
  })
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'This is the Message Returned'
  )
  expect(mockRouterPushFunction).toHaveBeenCalledTimes(1)
  expect(mockRouterPushFunction).toHaveBeenLastCalledWith(
    '/leagues/Test League'
  )
})

test('Create Event - Fails Validation', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  await wrapper.vm.createEvent()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )
})

test('Create Event - Successful Creation', async () => {
  axios.post.mockResolvedValue({ message: 'Event - X was Created' })
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const mockReturnToLeaguePage = jest.fn()
  wrapper.setMethods({ returnToLeaguePage: mockReturnToLeaguePage })
  wrapper.setData(sampleSingleEvent)
  await wrapper.vm.createEvent()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(1)
  expect(mockReturnToLeaguePage).toHaveBeenLastCalledWith({
    message: 'Event - X was Created',
  })
})

test('Create Event - Error in Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.post.mockRejectedValue({ message: 'Something went Wrong' })
  const mockReturnToLeaguePage = jest.fn()
  wrapper.setMethods('returnToLeaguePage', mockReturnToLeaguePage)
  wrapper.setData(sampleSingleEvent)
  await wrapper.vm.createEvent()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(0)
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Error: Problem Creating Event - Please Try Again'
  )
})

test('Create Event - Calls Correct API Location', async () => {
  axios.post.mockResolvedValue({ message: 'Event - X was Created' })
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData(sampleSingleEvent)
  sampleSingleEvent.id = undefined
  await wrapper.vm.createEvent()
  expect(axios.post).toHaveBeenCalledTimes(1)
  expect(axios.post).toHaveBeenLastCalledWith('/api/events', sampleSingleEvent)
  // Reset for Next Tests
  sampleSingleEvent.id = 'SprintelopeBushEstate2018-05-09'
})

test('Update Event - Calls Correct API Location', async () => {
  axios.put.mockRejectedValue({
    response: { data: { message: 'Event - X was Updated' } },
  })
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData(sampleSingleEvent)
  await wrapper.vm.updateEvent()
  const sampleSingleEventNoID = sampleSingleEvent
  sampleSingleEventNoID.id = undefined
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith(
    '/api/events/SprintelopeBushEstate2018-05-09',
    sampleSingleEventNoID
  )
})

test('Update Event - Fails Validation', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  await wrapper.vm.updateEvent()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )
})

test('Update Event - Successful Creation', async () => {
  axios.put.mockResolvedValue({ message: 'Event - X was Created' })
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const mockReturnToLeaguePage = jest.fn()
  wrapper.setMethods({ returnToLeaguePage: mockReturnToLeaguePage })
  wrapper.setData(sampleSingleEvent)
  await wrapper.vm.updateEvent()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(1)
  expect(mockReturnToLeaguePage).toHaveBeenLastCalledWith({
    message: 'Event - X was Created',
  })
})

test('Update Event - Error in Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  axios.put.mockRejectedValue({
    response: {
      data: { message: 'Error: Problem Updating Event - Please Try Again' },
    },
  })
  const wrapper = mount(EventForm, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const mockReturnToLeaguePage = jest.fn()
  wrapper.setMethods('returnToLeaguePage', mockReturnToLeaguePage)
  wrapper.setData(sampleSingleEvent)
  await wrapper.vm.updateEvent()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(0)
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Error: Problem Updating Event - Please Try Again'
  )
})
