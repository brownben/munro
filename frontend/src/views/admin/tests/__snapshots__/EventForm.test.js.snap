// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creation Form - Renders Correctly 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Create Event"
  />
   
  <div
    class="w-10/12 mx-auto md:w-3/4 xl:w-2/3 md:mt-2"
  >
    <h1
      class="my-2 text-3xl font-bold md:text-35xl font-heading text-main-700"
    >
      
    Create Event
  
    </h1>
     
    <div
      class="grid grid-cols-2 gap-8 pb-8 md:gap-6 lg:gap-8"
    >
      <form
        class="col-span-2"
      >
        <div>
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Name:"
          >
            Name:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Name:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Date: (YYYY-MM-DD)"
          >
            Date: (YYYY-MM-DD)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Date: (YYYY-MM-DD)"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Club/ Organiser:"
          >
            Club/ Organiser:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Club/ Organiser:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Website: (URL)"
          >
            Website: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Website: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Results: (URL)"
          >
            Results: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Results: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Winsplits: (URL)"
          >
            Winsplits: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Winsplits: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Routegadget: (URL)"
          >
            Routegadget: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Routegadget: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="League:"
          >
            
    League:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="League:"
          >
            <option />
             
          </select>
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="More Information:"
          >
            More Information:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="More Information:"
            type="text"
          />
        </div>
         
        <div
          class="mt-6 text-left"
        >
          <label
            class="m-0 mt-1 text-lg transition ease-out delay-500 select-none text-md font-heading text-main"
          >
            <span
              class="mr-3 align-top"
            >
              Allow User Submitted Results:
            </span>
             
            <input
              class="sr-only"
              type="checkbox"
            />
             
            <span
              class="inline-block w-4 h-4 mt-1 transition-all duration-300 ease-in-out border-2 border-main-300"
            />
          </label>
        </div>
         
        <button
          class="mt-8 button-lg"
        >
          Create Event
        </button>
         
        <!---->
      </form>
    </div>
  </div>
</div>
`;

exports[`Update Form - Renders Correctly 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Edit Event"
  />
   
  <div
    class="w-10/12 mx-auto md:w-3/4 xl:w-2/3 md:mt-2"
  >
    <h1
      class="my-2 text-3xl font-bold md:text-35xl font-heading text-main-700"
    >
      
    Edit Event
  
    </h1>
     
    <div
      class="grid grid-cols-2 gap-8 pb-8 md:gap-6 lg:gap-8"
    >
      <form
        class="col-span-2"
      >
        <div>
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Name:"
          >
            Name:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Name:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Date: (YYYY-MM-DD)"
          >
            Date: (YYYY-MM-DD)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Date: (YYYY-MM-DD)"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Club/ Organiser:"
          >
            Club/ Organiser:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Club/ Organiser:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Website: (URL)"
          >
            Website: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Website: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Results: (URL)"
          >
            Results: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Results: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Winsplits: (URL)"
          >
            Winsplits: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Winsplits: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Routegadget: (URL)"
          >
            Routegadget: (URL)
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Routegadget: (URL)"
            type="url"
          />
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="League:"
          >
            
    League:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="League:"
          >
            <option />
             
          </select>
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="More Information:"
          >
            More Information:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="More Information:"
            type="text"
          />
        </div>
         
        <div
          class="mt-6 text-left"
        >
          <label
            class="m-0 mt-1 text-lg transition ease-out delay-500 select-none text-md font-heading text-main"
          >
            <span
              class="mr-3 align-top"
            >
              Allow User Submitted Results:
            </span>
             
            <input
              class="sr-only"
              type="checkbox"
            />
             
            <span
              class="inline-block w-4 h-4 mt-1 transition-all duration-300 ease-in-out border-2 border-main-300"
            />
          </label>
        </div>
         
        <!---->
         
        <button
          class="mt-8 button-lg"
        >
          Update Event
        </button>
      </form>
    </div>
  </div>
</div>
`;
